File: Data Structures & Algorithms/target-sum/submission-10.java
class Solution {
    Map<String, Integer> map = new HashMap<>();
    private int sum = 0;
    
    // p = 正數 q = 負數絕對值
    // p + q = S
    // p - q = target
    // p  =  S + target / 2 ;
    // q  =  S - target / 2 
    public int findTargetSumWays(int[] nums, int target) {
        for (int num : nums) {
            sum += num;
        }
        sum -= Math.abs(target);
        if (sum < 0 || sum % 2 == 1) {
            return 0;
        }
        int n = nums.length;
        int capacity = sum / 2;
        int[][] memo = new int[n+1][capacity+1];
        for (int[] row : memo) {
            Arrays.fill(row, -1);
        }
        
        return dfs(n-1, capacity, nums, memo);
    }

    private int dfs(int i, int c, int[] nums, int[][] memo) {
        if (i < 0) {
            return c == 0 ? 1 : 0;
        }
        if (memo[i][c] != -1) {
            return memo[i][c];
        }
        int ans = dfs(i - 1, c, nums, memo);
        if (c >= nums[i]) {
            ans += dfs(i - 1, c - nums[i], nums, memo);
        }
        return memo[i][c] = ans;
    }
}

File: Data Structures & Algorithms/interleaving-string/submission-10.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        n, m = len(s1), len(s2)
        if n + m != len(s3):
            return False

        f = [False] * (m + 1)
        f[0] = True
        for j, y in enumerate(s2):
            f[j + 1] = f[j] and y == s3[j]
        for i, x in enumerate(s1):
            f[0] = f[0] and x == s3[i]
            for j, y in enumerate(s2):
                f[j + 1] = f[j + 1] and x == s3[i + j + 1] or \
                           f[j] and y == s3[i + j + 1]
        return f[m]


    #    f[i][j] = f[i-1][j] or f[i][j-1]
    #    f[i+1][j+1] = f[i][j+1] or f[i+1][j]


        # @cache  
        # def dfs(i: int, j: int) -> bool:
        #     if i < 0 and j < 0:
        #         return True
        #     return i >= 0 and s1[i] == s3[i + j + 1] and dfs(i - 1, j) or \
        #            j >= 0 and s2[j] == s3[i + j + 1] and dfs(i, j - 1)

        # return dfs(n - 1, m - 1)
